scripts/py_matter_idl/matter_idl/backwards_compatibility.py

```diff
@@ -0,0 +1,291 @@
+                f"Enumeration {cluster_name}::{original.name} switched base type from {original.base_type} to {updated.base_type}")
+
+        # Validate that all old entries exist
+        for entry in original.entries:
+            # old entry must exist and have identical code
+            existing = [item for item in updated.entries if item.name == entry.name]
+            if len(existing) == 0:
+                self._mark_incompatible(f"Enumeration {cluster_name}::{original.name} removed entry {entry.name}")
+            elif existing[0].code != entry.code:
+                self._mark_incompatible(
+                    f"Enumeration {cluster_name}::{original.name} changed code for entry {entry.name} from {entry.code} to {existing[0].code}")
+
+    def _check_bitmap_compatible(self, cluster_name: str, original: Bitmap, updated: Optional[Bitmap]):
+        if not updated:
+            self._mark_incompatible(f"Bitmap {cluster_name}::{original.name} was deleted")
+            return
+
+        if original.base_type != updated.base_type:
+            self._mark_incompatible(
+                f"Bitmap {cluster_name}::{original.name} switched base type from {original.base_type} to {updated.base_type}")
+
+        # Validate that all old entries exist
+        for entry in original.entries:
+            # old entry must exist and have identical code
+            existing = [item for item in updated.entries if item.name == entry.name]
+            if len(existing) == 0:
+                self._mark_incompatible(f"Bitmap {original.name} removed entry {entry.name}")
+            elif existing[0].code != entry.code:
+                self._mark_incompatible(
+                    f"Bitmap {original.name} changed code for entry {entry.name} from {entry.code} to {existing[0].code}")
+
+    def _check_event_compatible(self, cluster_name: str, event: Event, updated_event: Optional[Event]):
+        if not updated_event:
+            self._mark_incompatible(f"Event {cluster_name}::{event.name} was removed")
+            return
+
+        if event.code != updated_event.code:
+            self._mark_incompatible(f"Event {cluster_name}::{event.name} code changed from {event.code} to {updated_event.code}")
+
+        self._check_field_lists_are_the_same(f"Event {cluster_name}::{event.name}", event.fields, updated_event.fields)
+
+    def _check_command_compatible(self, cluster_name: str, command: Command, updated_command: Optional[Command]):
+        self.logger.debug(f"  Checking command {cluster_name}::{command.name}")
+        if not updated_command:
+            self._mark_incompatible(f"Command {cluster_name}::{command.name} was removed")
+            return
+
+        if command.code != updated_command.code:
+            self._mark_incompatible(
+                f"Command {cluster_name}::{command.name} code changed from {command.code} to {updated_command.code}")
+
+        if command.input_param != updated_command.input_param:
```

Comment:
Will this check for insertions of fields into a command/struct payloads that implicitly increment the field IDs of arguments downstream of that insertion?

Reply:
That would be checked part of "struct" compatibiltiy. Commands only have the input/output structure name, so changes in their content is not here but rather in the "struct compare".

Reply:
Regarding re-order and ID checks, this would detect it, however it will do the struct "allow renames". Specifically this is allowed:

```
OldRequest {
  int32u a = 1;
  int32u b = 2;
}

NewRequest {
  int32u b = 1;
  int32u a = 2;
}
```

Because it sees it is a rename. So if someone deletes "a" but adds "c" of the same type it would also pass. Was unsure what to do here, binary the structures are compatible but their meaning would be messed up.

Happy to change and not allow renames if we think that is useful. However we did historically have renames like "tempValue" becoming "temperatureValue" or similar.

Reply:
@bzbarsky-apple @mrjerryjohns - this is an interesting thing: I allowed renames because they seemed common. However this means if same data type, this would allow "add and remove at the same time" if the types align.

I would however claim that is uncommon and obvious in reviews of matter files ... I hope. If not I can always remove the "allow renames" logic.